File: pypy3/agent/q_learning.py
import agent.abs_agent as abs_agent
import random
import config
from typing import List
import q_table


class QLearning(abs_agent.Agent):
    """Q-learning agent"""

    def __init__(self, config: "config.Config"):
        super().__init__()
        self.alpha = float(config.cfg["AGENT_ALPHA"])
        self.gamma = float(config.cfg["AGENT_GAMMA"])
        self.eps = float(config.cfg["AGENT_EPSILON"])

    def a(self, q_table_: "q_table.QTable", s: int) -> int:
        """Eps-greedy policy."""

        if random.random() < self.eps:
            return random.randint(0, q_table_.a_space - 1)
        return self.argmax(q_table_.table[s])

    def learn(self, q_table_: "q_table.QTable", s1: int, a1: int, r: float, s2: int, a2: int):
        """Learn."""
        q_table_.table[s1][a1] = (1. - self.alpha) * q_table_.table[s1][a1] + \
            self.alpha * (r + self.gamma * max(q_table_.table[s2]))

    def fix(self):
        """Change parameters not to learn."""
        self.alpha = 0.
        self.eps = 0.

    def argmax(self, list_: List[int]):
        """Argmax."""
        res = 0
        for i, q in enumerate(list_):
            if q > list_[res]:
                res = i
        return res

File: Rust_rustc/src/agent/q_learning.rs
use crate::agent;
use crate::config::Config;
use crate::q_table::QTable;
use crate::{Action, Reward, State};
use rand;
use rand::prelude::*;
use std::error::Error;
use std::f64;

pub struct QLearning {
    alpha: f64,
    gamma: f64,
    eps: f64,

    rng: ThreadRng,
}

impl QLearning {
    pub fn new(config: &Config) -> Result<QLearning, Box<dyn Error>> {
        let alpha = config.get("AGENT_ALPHA")?.parse()?;
        let gamma = config.get("AGENT_GAMMA")?.parse()?;
        let eps: f64 = config.get("AGENT_EPSILON")?.parse()?;
        let rng = rand::thread_rng();
        Ok(QLearning {
            alpha,
            gamma,
            eps,
            rng,
        })
    }
}

impl agent::Agent for QLearning {
    fn action(&mut self, q_table: &QTable, s: State) -> Action {
        if self.rng.gen::<f64>() < self.eps {
            self.rng.gen_range(0, q_table.action_size)
        } else {
            argmax(&q_table.table[s])
        }
    }

    fn learn(
        &self,
        q_table: &mut QTable,
        s1: State,
        a1: Action,
        r: Reward,
        s2: State,
        _a2: Action,
    ) {
        let max = q_table.table[s2]
            .iter()
            .fold(f64::NEG_INFINITY, |m, &v| m.max(v));

        let alpha = self.alpha;
        let gamma = self.gamma;

        q_table.table[s1][a1] = (1.0 - alpha) * q_table.table[s1][a1] + alpha * (r + gamma * max);
    }

    fn fix(&mut self) {
        self.alpha = 0.;
        self.eps = 0.;
    }
}

fn argmax<T: PartialOrd>(slice: &[T]) -> usize {
    let mut res = 0;
    for i in 1..slice.len() {
        if slice[res] < slice[i] {
            res = i
        }
    }
    return res;
}
